File: python-impl/ui/bank_app_ui.py
from datetime import datetime
import re

from tabulate import tabulate
from insfrastructure.entities.bank_account_entity import BankAccountEntity
from insfrastructure.entities.person_entity import PersonEntity
from insfrastructure.enums.account_types import AccountType
from insfrastructure.enums.gender_choice import Gender
from insfrastructure.validators.input_validator import InputValidator


class BankAppUI:
    @staticmethod
    def display_menu():
        single_dashes = "-" * 50
        double_dashes = "=" * 50
        print("\n")
        print(double_dashes)
        print("\tWelcome to the Bank App CLI")
        print(single_dashes)
        print("\nPlease select an option from the menu below")

        print("\n")
        print("1. Create a new account")
        print("2. Display all accounts")
        print("3. Update an account")
        print("4. Delete an account")
        print("5. Deposit an amount into your account")
        print("6. Withdraw an amount from your account")
        print("7. Search for account")
        print("8. Exit")
        print("\n")

    @staticmethod
    def get_user_input( message: str = "Enter your choice: ", returnNoneOnEmpty: bool = False):
        while True:
            input_message = input(message)
            if not input_message.strip():
                if returnNoneOnEmpty: 
                    return None
                else:
                    continue
            return input_message
        
    @staticmethod
    def show_user_details( person: PersonEntity):
        print("Showing account details...\n")
        headers = ["Attribute", "Value"]
        data = [
            ["Name", f"{person.first_name} {person.last_name}"],
            ["Account holder's email", person.email],
            ["Account holder's phone number", person.phone_number],
            ["Account holder's national ID", person.national_id],
            ["Account holder's date of birth", person.date_of_birth],
            ["Account holder's gender", person.gender],
            ["Account holder's permanent address", person.parmanent_address]
            
        ]
        table = tabulate(data, headers=headers, tablefmt="pretty")
        print(table)
    
    @staticmethod
    def show_account_details(account: BankAccountEntity, details_text: bool = True):
        if details_text:
            print("Showing account details...\n")
        headers = ["Attribute", "Value"]
        data = [
            ["Account number", account.account_number],
            ["Account holder's name", f"{account.account_holder.first_name} {account.account_holder.last_name}"],
            ["Balance", account.balance],
            ["Account type", account.account_type],
            ["Creation date", account.creation_date],
            ["Account holder's email", account.account_holder.email],
            ["Account holder's phone number", account.account_holder.phone_number],
            ["Account holder's national ID", account.account_holder.national_id],
            ["Account holder's date of birth", account.account_holder.date_of_birth],
            ["Account holder's gender", account.account_holder.gender],
            ["Account holder's permanent address", account.account_holder.parmanent_address]
            
        ]
        table = tabulate(data, headers=headers, tablefmt="pretty")
        print(table)
    
    @staticmethod
    def show_account_types(minimum_initial_deposit, minimum_balance_before_withdrawal):
        print("Account types we offer...")
        headers = ["Account Type", "Minimum Initial Deposit", "Minimum Balance Before Withdrawal"]
        data = []
        for account_type in AccountType:
            data.append([account_type.name, minimum_initial_deposit[account_type], minimum_balance_before_withdrawal[account_type]])
        table = tabulate(data, headers=headers, tablefmt="pretty")
        print(table)

    @staticmethod
    def display_all_accounts(accounts):
        print("Displaying all accounts...\n")
        if not accounts:
            print("No accounts found.")
        for account in accounts:
            BankAppUI.show_account_details(account, details_text=False)

    @staticmethod
    def get_valid_gender(prompt, returnNoneOnEmpty: bool = False):
        while True:
            gender_input = BankAppUI.get_user_input(prompt, returnNoneOnEmpty=True)
            
            if gender_input is None:
                if returnNoneOnEmpty:
                    return None
            else:
                gender = InputValidator.validate_gender(gender_input)
                if gender is not None:
                    return gender
            
                print("Invalid gender choice. Please enter \"male\", \"female\", or \"other\".")

    @staticmethod
    def get_valid_date(prompt, returnNoneOnEmpty: bool = False):
        while True:
            date_str = BankAppUI.get_user_input(prompt, returnNoneOnEmpty=True)
            if date_str is None:
                if returnNoneOnEmpty:
                    return None
            else:
                date = InputValidator.validate_date(date_str)
                if date is not None:
                    return date
            print("Invalid date format. Please enter date in the format YYYY-MM-DD.")


    @staticmethod
    def get_valid_email(prompt, returnNoneOnEmpty: bool = False):
        while True:
            email = BankAppUI.get_user_input(prompt, returnNoneOnEmpty=True)
            if email is None:
                if returnNoneOnEmpty:
                    return None
            else:
                validated_email = InputValidator.validate_email(email)
                if validated_email is not None:
                    return validated_email
                print("Invalid email address format. Please enter a valid email address.")


    @staticmethod
    def get_valid_amount(prompt, minimum_amount=None, account_type=None, returnNoneOnEmpty: bool = False):
        while True:
            amount = BankAppUI.get_user_input(prompt, returnNoneOnEmpty=True)
            if amount is None:
                if returnNoneOnEmpty:
                    return None
            else:
                validated_amount = InputValidator.validate_amount(amount)
                if validated_amount is not None:
                    if minimum_amount and account_type:
                        if validated_amount < minimum_amount[account_type]:
                            print(f"Amount must be greater than {minimum_amount[account_type]} for {account_type.name} account. Please enter a valid amount.")
                            continue
                    return validated_amount
                print("Invalid amount. Please enter a valid amount.")
                


    @staticmethod
    def get_valid_withdraw_amount(prompt, minimum_withdraw_amount, account, returnNoneOnEmpty: bool = False):
        while True:
            amount = BankAppUI.get_user_input(prompt, returnNoneOnEmpty=True)
            if amount is None:
                if returnNoneOnEmpty:
                    return None
            else:
                validated_amount = InputValidator.validate_withdraw_amount(amount, minimum_withdraw_amount[account.account_type], account.balance if account else 0)
                if validated_amount is not None:
                    return validated_amount
                print(f"Invalid amount. Account must have {minimum_withdraw_amount[account.account_type]} for a {account.account_type.name} account after withdrawal. Please enter a valid amount.")


    @staticmethod
    def get_valid_account_type(prompt, minimum_initial_deposit, minimum_balance_before_withdrawal, returnNoneOnEmpty: bool = False):
        BankAppUI.show_account_types(minimum_initial_deposit=minimum_initial_deposit, minimum_balance_before_withdrawal=minimum_balance_before_withdrawal)
        while True:
            account_type_input = BankAppUI.get_user_input(prompt, returnNoneOnEmpty=True)
            if account_type_input is None:
                if returnNoneOnEmpty:
                    return None
            else:
                validated_account_type = InputValidator.validate_account_type(account_type_input)
                if validated_account_type is not None:
                    return validated_account_type
                print("Invalid account type. Please enter either one of the following: ")
                for account_type in AccountType:
                    print(account_type.name.lower())


    @staticmethod
    def get_valid_account_id(prompt, returnNoneOnEmpty: bool = False):
        while True:
            account_id = BankAppUI.get_user_input(prompt, returnNoneOnEmpty=True)
            if account_id is None:
                return None
            else:
                validated_account_id = InputValidator.validate_account_id(account_id)
                if validated_account_id is not None:
                    return validated_account_id
                print("Invalid account number. Please enter a valid account number.")

File: typescript-impl/src/ui/bank_app_ui.ts
import { InputValidator } from '../infrastructure/validators/input_validator'; // Import the InputValidator class from its file
import { BankAccountEntity } from '../infrastructure/entities/bank_account_entity'; // Import the BankAccountEntity class from its file
import { PersonEntity } from '../infrastructure/entities/person_entity'; // Import the PersonEntity class from its file
import { AccountType } from '../infrastructure/enums/account_type'; // Import the AccountType enum from its file
import { Gender } from '../infrastructure/enums/gender_choice';
import { table } from 'console';

const readline = require('readline');

export class BankAppUI {
    display_menu(): void {
        const singleDashes: string = "-".repeat(50);
        const doubleDashes: string = "=".repeat(50);
        console.log("\n");
        console.log(doubleDashes);
        console.log("\tWelcome to the Bank App CLI");
        console.log(singleDashes);
        console.log("\nPlease select an option from the menu below");

        console.log("\n");
        console.log("1. Create a new account");
        console.log("2. Display all accounts");
        console.log("3. Update an account");
        console.log("4. Delete an account");
        console.log("5. Deposit an amount into your account");
        console.log("6. Withdraw an amount from your account");
        console.log("7. Search for account");
        console.log("8. Exit");
        console.log("\n");
    }

    async get_user_input(message: string = "Enter your choice: ", returnNoneOnEmpty: boolean = false): Promise<string | null> {
        while (true) {
            const inputMessage: string | null = await this.prompt_user(message) || "";
            if (inputMessage.length === 0) { // Check if the input string is empty
                if (returnNoneOnEmpty) {
                    return null;
                } else {
                    continue;
                }
            }
            return inputMessage;
        }
    }

    private prompt_user(message: string) {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });

        return new Promise<string>((resolve) => {
            rl.question(message, (input: string) => {
                rl.close();
                resolve(input);
            });
        });
    }

    show_user_details(person: PersonEntity): void {
        console.log("Showing account details...\n");
        const headers: string[] = ["Attribute", "Value"];
        const data: any[][] = [
            ["Name", `${person.first_name} ${person.last_name}`],
            ["Account holder's email", person.email],
            ["Account holder's phone number", person.phone_number],
            ["Account holder's national ID", person.national_id],
            ["Account holder's date of birth", person.date_of_birth],
            ["Account holder's gender", person.gender],
            ["Account holder's permanent address", person.permanent_address]
        ];
        // const table: string = this.formatTable(data, headers);
        // console.log(table(data));
        this.formatTable(data);
    }

    show_account_details(account: BankAccountEntity, detailsText: boolean = true): void {
        if (detailsText) {
            console.log("Showing account details...\n");
        }
        const headers: string[] = ["Attribute", "Value"];
        const data: any[][] = [
            ["Account number", account.account_number],
            ["Account holder's name", `${account.account_holder.first_name} ${account.account_holder.last_name}`],
            ["Balance", account.balance],
            ["Account type", account.account_type],
            ["Creation date", account.creation_date],
            ["Account holder's email", account.account_holder.email],
            ["Account holder's phone number", account.account_holder.phone_number],
            ["Account holder's national ID", account.account_holder.national_id],
            ["Account holder's date of birth", account.account_holder.date_of_birth],
            ["Account holder's gender", account.account_holder.gender],
            ["Account holder's permanent address", account.account_holder.permanent_address]
        ];
        // const table: string = this.formatTable(data, headers);
        // console.log(table(data));
        this.formatTable(data);
    }

    show_account_types(minimumInitialDeposit: Map<AccountType, number>, minimumBalanceBeforeWithdrawal: Map<AccountType, number>): void {
        console.log("Account types we offer...");
        const headers: string[] = ["Account Type", "Minimum Initial Deposit", "Minimum Balance Before Withdrawal"];
        const data: any[][] = [];
        for (const [accountType, initialDeposit] of minimumInitialDeposit.entries()) {
            data.push([accountType, initialDeposit, minimumBalanceBeforeWithdrawal.get(accountType)]);
        }
        // const table: string = this.formatTable(data, headers);
        // console.log(table(data));
        // table(data);
        this.formatTable(data);
    }

    display_all_accounts(accounts: BankAccountEntity[]): void {
        console.log("Displaying all accounts...\n");
        if (!accounts.length) {
            console.log("No accounts found.");
        }
        for (const account of accounts) {
            this.show_account_details(account, false);
        }
    }

    async get_valid_gender(prompt: string, returnNoneOnEmpty: boolean = false): Promise<Gender | null> {
        while (true) {
            const genderInput: string | null =await this.get_user_input(prompt, returnNoneOnEmpty);
            if (genderInput === null) {
                return null;
            }
            const gender: Gender | null = InputValidator.validateGender(genderInput);
            if (gender !== null) {
                return gender;
            }
            console.log("Invalid gender choice. Please enter \"male\", \"female\", or \"other\".");
        }
    }

    async get_valid_date(prompt: string, returnNoneOnEmpty: boolean = false): Promise<Date | null> {
        while (true) {
            const dateStr: string | null =await this.get_user_input(prompt, returnNoneOnEmpty);
            if (dateStr === null) {
                return null;
            }
            const date: Date | null = InputValidator.validateDate(dateStr);
            if (date !== null) {
                return date;
            }
            console.log("Invalid date format. Please enter date in the format YYYY-MM-DD.");
        }
    }

    async get_valid_email(prompt: string, returnNoneOnEmpty: boolean = false): Promise<string | null> {
        while (true) {
            const email: string | null =await this.get_user_input(prompt, returnNoneOnEmpty);
            if (email === null) {
                return null;
            }
            const validatedEmail: string | null = InputValidator.validateEmail(email);
            if (validatedEmail !== null) {
                return validatedEmail;
            }
            console.log("Invalid email address format. Please enter a valid email address.");
        }
    }

    async get_valid_amount(prompt: string, minimumAmount?: Map<AccountType, number>, accountType?: AccountType, returnNoneOnEmpty: boolean = false): Promise<number | null> {
        while (true) {
            const amount: string | null =await this.get_user_input(prompt, returnNoneOnEmpty);
            if (amount === null) {
                return null;
            }
            const validatedAmount: number | null = InputValidator.validateAmount(amount);
            if (validatedAmount !== null) {
                if (minimumAmount && accountType) {
                    if (validatedAmount < (minimumAmount.get(accountType) || 0)) {
                        console.log(`Amount must be greater than ${minimumAmount.get(accountType)} for ${accountType} account. Please enter a valid amount.`);
                        continue;
                    }
                }
                return validatedAmount;
            }
            console.log("Invalid amount. Please enter a valid amount.");
        }
    }

    async get_valid_withdraw_amount(prompt: string, minimumWithdrawAmount: Map<AccountType, number>, account: BankAccountEntity, returnNoneOnEmpty: boolean = false): Promise<number | null> {
        while (true) {
            const amount: string | null =await this.get_user_input(prompt, returnNoneOnEmpty);
            if (amount === null) {
                return null;
            }
            const validatedAmount: number | null = InputValidator.validateWithdrawAmount(amount, minimumWithdrawAmount.get(account.account_type) || 0, account.balance || 0);
            if (validatedAmount !== null) {
                return validatedAmount;
            }
            console.log(`Invalid amount. Account must have ${minimumWithdrawAmount} for a ${account.account_type} account after withdrawal. Please enter a valid amount.`);
        }
    }

    async get_valid_account_type(prompt: string, minimumInitialDeposit: Map<AccountType, number>, minimumBalanceBeforeWithdrawal: Map<AccountType, number>, returnNoneOnEmpty: boolean = false): Promise<AccountType | null> {
        this.show_account_types(minimumInitialDeposit, minimumBalanceBeforeWithdrawal);
        while (true) {
            const accountTypeInput: string | null =await this.get_user_input(prompt, returnNoneOnEmpty);
            if (accountTypeInput === null) {
                return null;
            }
            const validatedAccountType: AccountType | null = InputValidator.validateAccountType(accountTypeInput);
            if (validatedAccountType !== null) {
                return validatedAccountType;
            }
            console.log("Invalid account type. Please enter either one of the following: ");

            for (const accountType in AccountType) {
                console.log(accountType);
            }
        }
    }

    async get_valid_account_id(prompt: string, returnNoneOnEmpty: boolean = false): Promise<number | null> {
        while (true) {
            const accountId: string | null =await this.get_user_input(prompt, returnNoneOnEmpty);
            if (accountId === null) {
                return null;
            }
            const validatedAccountId: number | null = InputValidator.validateAccountId(accountId);
            if (validatedAccountId !== null) {
                return validatedAccountId;
            }
            console.log("Invalid account number. Please enter a valid account number.");
        }
    }

    private formatTable(data: any[][]): void {
        table(data);
    }
}
